File: xhy_paper_consult/proxy.py
import http.server
import threading
import urllib.request
import json
import os

PORT = 3000
DIFY_URL = 'https://api.dify.ai/v1/workflows/run'
API_KEY = 'app-r2ny0MjklS9YYhBRvN80j3S9'

class ProxyHandler(http.server.BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != '/api/proxy':
            self.send_error(404)
            return

        length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(length).decode('utf-8')

        req = urllib.request.Request(DIFY_URL, data=body.encode('utf-8'))
        req.add_header('Authorization', f'Bearer {API_KEY}')
        req.add_header('Content-Type', 'application/json')
        req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
        req.add_header('Accept', '*/*')

        try:
            with urllib.request.urlopen(req, timeout=120) as upstream:
                self.send_response(upstream.status)
                self.send_header('Content-Type', 'text/plain; charset=utf-8')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Cache-Control', 'no-cache')
                self.end_headers()
                data = upstream.read()
                # 收集所有 SSE 行
                lines = data.decode('utf-8').split('\n')
                results = []
                for line in lines:
                    if line.startswith('data:'):
                        try:
                            chunk = json.loads(line[5:].strip())
                            ev = chunk.get('event', '')
                            if ev == 'workflow_finished':
                                outputs = chunk.get('data', {}).get('outputs', {})
                                results = outputs.get('results', [])
                        except json.JSONDecodeError:
                            pass
                # 返回简化结果
                resp_data = json.dumps({'results': results}, ensure_ascii=False)
                self.wfile.write(resp_data.encode('utf-8'))
        except urllib.error.HTTPError as e:
            err = e.read()
            self.send_response(e.code)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(err)
        except Exception as e:
            msg = json.dumps({'results': [], 'error': str(e)}).encode()
            self.send_response(502)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(msg)

    def do_GET(self):
        # 提供静态文件
        if self.path == '/' or self.path == '/index.html':
            self.path = '/index.html'
        try:
            filepath = os.path.join(os.path.dirname(os.path.abspath(__file__)), self.path.lstrip('/'))
            if os.path.isfile(filepath):
                with open(filepath, 'rb') as f:
                    content = f.read()
                ct = 'text/html' if filepath.endswith('.html') else 'text/plain'
                self.send_response(200)
                self.send_header('Content-Type', ct)
                self.end_headers()
                self.wfile.write(content)
            else:
                self.send_error(404)
        except Exception:
            self.send_error(500)

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def log_message(self, format, *args):
        pass  # 静默日志

class ThreadedHTTPServer(http.server.ThreadingHTTPServer):
    """多线程服务器，支持并发请求"""
    daemon_threads = True

os.chdir(os.path.dirname(os.path.abspath(__file__)))
print(f'Server: http://localhost:{PORT}')
ThreadedHTTPServer(('', PORT), ProxyHandler).serve_forever()

File: xhy_aichat/api/proxy.js
/**
 * Vercel Serverless Function - Dify API 代理
 * 
 * 解决 Vercel 静态托管下 /proxy/* 请求的转发问题。
 * 通过 vercel.json rewrite 将 /proxy/* 映射到此函数。
 * 
 * 需在 Vercel 环境变量中配置：
 *   DIFY_API_KEY  - 你的 Dify API Key（必需）
 *   DIFY_API_BASE - Dify API 地址（可选，默认 https://api.dify.ai/v1）
 */
export default async function handler(req, res) {
  // CORS 预检
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    return res.status(200).end();
  }

  const apiKey = process.env.DIFY_API_KEY;
  if (!apiKey) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    return res.status(500).json({ error: 'DIFY_API_KEY 未配置，请在 Vercel 环境变量中设置' });
  }

  const apiBase = process.env.DIFY_API_BASE || 'https://api.dify.ai/v1';

  // 从 rewrite 规则中获取原始路径
  const difyPath = req.query.dify_path || '';

  // 重构查询参数（排除 dify_path）
  const url = new URL(req.url, 'http://localhost');
  url.searchParams.delete('dify_path');
  const queryString = url.search;

  const targetUrl = `${apiBase}/${difyPath}${queryString}`;

  // 读取请求体
  let body = '';
  if (req.method !== 'GET' && req.method !== 'HEAD') {
    const chunks = [];
    for await (const chunk of req) {
      chunks.push(chunk);
    }
    body = Buffer.concat(chunks).toString();
  }

  try {
    const fetchResp = await fetch(targetUrl, {
      method: req.method,
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': 'Vercel-Dify-Proxy/1.0',
      },
      body: body || undefined,
      signal: AbortSignal.timeout(120000), // 2 分钟超时
    });

    const contentType = fetchResp.headers.get('content-type') || 'application/json';
    const responseBody = await fetchResp.text();

    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Content-Type', contentType);
    return res.status(fetchResp.status).send(responseBody);
  } catch (error) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    if (error.name === 'TimeoutError' || error.name === 'AbortError') {
      return res.status(504).json({ error: 'Dify API 请求超时' });
    }
    return res.status(502).json({ error: '代理请求失败: ' + error.message });
  }
}
